### DISCUSSION TITLE
GPU training and pipe-line training from scratch

### DISCUSSION THREAD

**Discussion by vijaysumaravi** [pyannote/pyannote-audio#595] [Q&A]
Hi,

I have two questions related to 'pyannote-pipeline'

1. Fine-tuning pipelines on a GPU. 
I am trying to train the pretrained speaker diarization pipelines on my own dataset. I am running this command from the tutorial: 
`pyannote-pipeline train --subset=development --iterations=50 ${EXP_DIR} my.personal.dataset`

However, GPU usage is still 0%. The fine-tuning is happening on CPUs instead. I tried providing the CUDA_AVAILABLE_DEVICES but that doesn't help either. 
The documentation says GPU is used by default when available but I don't see that happening. I see you've mentioned that the latest branch solves this issue but I have the latest dev branch and I am still facing this.  Am I doing something wrong?

2. How do I train the whole pipeline from scratch? The documentation has training details for individual blocks. Is that the only way - train individual blocks separately, if I don't want to use pretrained models? 

I was not sure of opening issues for these so I started a discussion thread instead. Thank you!

**Comment by hbredin** [comment]
> However, GPU usage is still 0%. The fine-tuning is happening on CPUs instead. I tried providing the CUDA_AVAILABLE_DEVICES but that doesn't help either.

`pyannote.pipeline` does not run on GPU for the simple reason that it assumes that all neural network stuff has happened before (when precomputing VAD  scores or embeddings.).

> The documentation says GPU is used by default when available 
Does it? Can you please provide a link?

>     1. How do I train the whole pipeline from scratch? The documentation has training details for individual blocks. Is that the only way - train individual blocks separately, if I don't want to use pretrained models?

Until a really end-to-end diarization is implemented, that is indeed the only way: 
- first train individual models 
- then train the diarization pipeline on top of those
 
> I was not sure of opening issues for these so I started a discussion thread instead. Thank you!
I think you made the right choice :)
